File: python/0003.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        t = dict()
        ans = 0
        beg = 0
        for i in range(len(s)):
            if s[i] in t and t[s[i]] >= beg:
                beg = t[s[i]] + 1
            t[s[i]] = i
            ans = max(ans, i - beg + 1)
            # print(beg,i,s[i],ans,t)
        return ans


if __name__ == '__main__':
    s = Solution()
    ans = s.lengthOfLongestSubstring("abcabcbb")
    print(ans)

File: rust/src/bin/0003.rs
use std::collections::{HashMap};

struct Solution;

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        if s.len() == 0 {
            return 0;
        }
        let mut a = HashMap::<char, usize>::new();
        let mut i = 0usize;
        let mut ans = 0usize;
        let chars: Vec<char> = s.chars().collect();
        for j in 0..chars.len() {
            let c = &chars[j];
            if a.contains_key(c) {
                let pos = *a.get(c).unwrap();
                while i <= pos {
                    a.remove(&chars[i]);
                    i += 1;
                }
            }
            a.insert(*c, j);
            let len = j - i + 1;
            // println!("s[{}]={} len={} i={}", j, c, len, i);
            if ans < len {
                ans = len;
            }
        }
        return ans as i32;
    }
}

fn main() {
    let ans = Solution::length_of_longest_substring("a".to_string());
    println!("{}", ans);
}